build.sh

```diff
@@ -1,73 +1,77 @@
+set -e
 
-function setuptree() {
-  b=$( mktemp -d $buildpath/gh-ostXXXXXX ) || return 1
-  mkdir -p $b/gh-ost
-  mkdir -p $b/gh-ost/usr/bin
-  echo $b
-}
-
-function build {
-  osname=$1
-  osshort=$2
-  GOOS=$3
-  GOARCH=$4
+. ./script/common
+buildpath=build
+timestamp=$(date "+%Y%m%d%H%M%S")
 
+function build() {
+  goos=$1
+  osshort=$goos
 
+  envs="-e GOOS=$goos \
+    -e GOARCH=amd64 \
+    -e CGO_ENABLED=1 \
+    -e GOCACHE=/go/src/${PACKAGE_PATH}/.cache \
+    -e RELEASE_VERSION=${RELEASE_VERSION} \
+    -e BIN_PATH=${buildpath}/bin"
 
-  if ! go version | egrep -q 'go(1[.]9|1[.]1[0-9])' ; then
-    echo "go version is too low. Must use 1.9 or above"
-    exit 1
+  if [ "$goos" == "darwin" ]; then
+    # specify cross compiler
+    envs="$envs -e CC=x86_64h-apple-darwin14-cc"
+    osshort="osx"
   fi
 
-  echo "Building ${osname} binary"
-  export GOOS
-  export GOARCH
-  go build -ldflags "$ldflags" -o $buildpath/$target go/cmd/gh-ost/main.go
-
-  if [ $? -ne 0 ]; then
-      echo "Build failed for ${osname}"
-      exit 1
-  fi
+  echo "Building $goos binary in ${buildpath}/bin"
 
-  (cd $buildpath && tar cfz ./gh-ost-binary-${osshort}-${timestamp}.tar.gz $target)
+  docker run -u $(id -u):$(id -g) --rm $envs \
+    -v $(pwd):/go/src/${PACKAGE_PATH} \
+    -w /go/src/${PACKAGE_PATH} \
+    dockercore/golang-cross:$GO_VERSION \
+    bash -c 'go build -ldflags "-X main.AppVersion=${RELEASE_VERSION}" -o ${BIN_PATH}/gh-ost-$GOOS-$GOARCH go/cmd/gh-ost/main.go'
 
-  if [ "$GOOS" == "linux" ] ; then
-    echo "Creating Distro full packages"
-    builddir=$(setuptree)
-    cp $buildpath/$target $builddir/gh-ost/usr/bin
-    cd $buildpath
-    fpm -v "${RELEASE_VERSION}" --epoch 1 -f -s dir -n gh-ost -m 'shlomi-noach <shlomi-noach+gh-ost-deb@github.com>' --description "GitHub's Online Schema Migrations for MySQL " --url "https://github.com/github/gh-ost" --vendor "GitHub" --license "Apache 2.0" -C $builddir/gh-ost --prefix=/ -t rpm .
-    fpm -v "${RELEASE_VERSION}" --epoch 1 -f -s dir -n gh-ost -m 'shlomi-noach <shlomi-noach+gh-ost-deb@github.com>' --description "GitHub's Online Schema Migrations for MySQL " --url "https://github.com/github/gh-ost" --vendor "GitHub" --license "Apache 2.0" -C $builddir/gh-ost --prefix=/ -t deb --deb-no-default-config-files .
-  fi
+  (
+    cp ${buildpath}/bin/gh-ost-$goos-amd64 $buildpath/dist/gh-ost \
+    && cd $buildpath/dist \
+    && tar cfz ./gh-ost-binary-${osshort}-${timestamp}.tar.gz gh-ost\
+    && rm gh-ost
+  )
 }
 
-main() {
-  if [ -z "${RELEASE_VERSION}" ] ; then
-    RELEASE_VERSION=$(git describe --abbrev=0 --tags | tr -d 'v')
-  fi
-  if [ -z "${RELEASE_VERSION}" ] ; then
-    RELEASE_VERSION=$(cat RELEASE_VERSION)
-  fi
-
-
-  buildpath=/tmp/gh-ost-release
-  target=gh-ost
-  timestamp=$(date "+%Y%m%d%H%M%S")
-  ldflags="-X main.AppVersion=${RELEASE_VERSION}"
+function distro_packages() {
+  echo "Creating Distro full packages"
+  docker build -t gh-ost/fpm -f docker/Dockerfile.fpm docker
```

Comment:
It's true that this will build the `fpm` image every time, but the docker layer cache will make this super fast on repeated runs.

Reply:
At some point I will want to get rid of `fpm`.